File: module-04-ip-addressing-subnetting/subnet_quiz.py
"""
Module 04 -- an infinite subnetting drill generator. Every problem is
computed from subnet_calc.py's own logic, so the "answer key" is never
hand-typed anywhere -- it's derived the same way your own answers should be.

Modes:
  python subnet_quiz.py            interactive quiz, 10 questions
  python subnet_quiz.py --auto 5   print 5 generated problems + answers,
                                    no input() required (used to verify the
                                    generator itself produces sane problems --
                                    see DECISIONS.md for why this mode exists)
"""
import random
import sys

from subnet_calc import subnet_info

QUESTION_TYPES = ["network", "broadcast", "first_host", "last_host", "usable_hosts", "netmask"]

QUESTION_TEXT = {
    "network": "What is the NETWORK address of {cidr}?",
    "broadcast": "What is the BROADCAST address of {cidr}?",
    "first_host": "What is the FIRST usable host address in {cidr}?",
    "last_host": "What is the LAST usable host address in {cidr}?",
    "usable_hosts": "How many USABLE HOST addresses does {cidr} have?",
    "netmask": "What is the dotted-decimal NETMASK for {cidr}?",
}

FIELD_FOR_TYPE = {
    "network": "network",
    "broadcast": "broadcast",
    "first_host": "first_host",
    "last_host": "last_host",
    "usable_hosts": "usable_hosts",
    "netmask": "netmask",
}


def random_cidr(rng: random.Random) -> str:
    # Keep prefixes in a range where "usable hosts" questions have a
    # pedagogically interesting, non-trivial answer (not /31 or /32 --
    # subnet_calc.py already covers those edge cases in its own docstring).
    octet1 = rng.choice([10, 172, 192])
    octet2 = rng.randint(0, 255) if octet1 != 172 else rng.randint(16, 31)
    octet3 = rng.randint(0, 255)
    prefix = rng.randint(24, 29)
    # zero out the host bits so the CIDR you're given is already a valid
    # network address, like a real exam question would present it
    ip = f"{octet1}.{octet2}.{octet3}.0"
    info = subnet_info(f"{ip}/{prefix}")
    return f"{info['network']}/{prefix}"


def generate_problem(rng: random.Random) -> dict:
    cidr = random_cidr(rng)
    qtype = rng.choice(QUESTION_TYPES)
    info = subnet_info(cidr)
    answer = info[FIELD_FOR_TYPE[qtype]]
    return {
        "cidr": cidr,
        "qtype": qtype,
        "question": QUESTION_TEXT[qtype].format(cidr=cidr),
        "answer": str(answer),
    }


def run_interactive(num_questions=10, rng=None):
    rng = rng or random.Random()
    score = 0
    for i in range(1, num_questions + 1):
        problem = generate_problem(rng)
        print(f"\nQ{i}. {problem['question']}")
        given = input("> ").strip()
        if given == problem["answer"]:
            print("Correct.")
            score += 1
        else:
            print(f"Incorrect. Correct answer: {problem['answer']}")
    print(f"\nFinal score: {score}/{num_questions}")
    return score


def run_auto(num_questions=5, rng=None):
    """Non-interactive: prints generated problems + answers so the
    generator's own correctness can be checked without a human typing
    answers (used during development/testing of this file itself)."""
    rng = rng or random.Random()
    for i in range(1, num_questions + 1):
        problem = generate_problem(rng)
        print(f"Q{i}. {problem['question']}")
        print(f"    Answer: {problem['answer']}")


if __name__ == "__main__":
    if len(sys.argv) >= 2 and sys.argv[1] == "--auto":
        n = int(sys.argv[2]) if len(sys.argv) >= 3 else 5
        run_auto(n)
    else:
        n = int(sys.argv[1]) if len(sys.argv) >= 2 else 10
        run_interactive(n)

File: module-04-ip-addressing-subnetting/subnet_quiz.js
/**
 * Module 04 — JS/Node port of subnet_quiz.py. Same generator logic,
 * built on subnet_calc.js's subnetInfo() as the answer key.
 *
 * Modes:
 *   node subnet_quiz.js            interactive quiz, 10 questions
 *   node subnet_quiz.js --auto 5   non-interactive, prints problems+answers
 */
const readline = require("readline");
const { subnetInfo } = require("./subnet_calc.js");

const QUESTION_TYPES = ["network", "broadcast", "firstHost", "lastHost", "usableHosts", "netmask"];

const QUESTION_TEXT = {
  network: (cidr) => `What is the NETWORK address of ${cidr}?`,
  broadcast: (cidr) => `What is the BROADCAST address of ${cidr}?`,
  firstHost: (cidr) => `What is the FIRST usable host address in ${cidr}?`,
  lastHost: (cidr) => `What is the LAST usable host address in ${cidr}?`,
  usableHosts: (cidr) => `How many USABLE HOST addresses does ${cidr} have?`,
  netmask: (cidr) => `What is the dotted-decimal NETMASK for ${cidr}?`,
};

function randInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function randomCidr() {
  const first = [10, 172, 192][randInt(0, 2)];
  const second = first === 172 ? randInt(16, 31) : randInt(0, 255);
  const third = randInt(0, 255);
  const prefix = randInt(24, 29);
  const ip = `${first}.${second}.${third}.0`;
  const info = subnetInfo(`${ip}/${prefix}`);
  return `${info.network}/${prefix}`;
}

function generateProblem() {
  const cidr = randomCidr();
  const qtype = QUESTION_TYPES[randInt(0, QUESTION_TYPES.length - 1)];
  const info = subnetInfo(cidr);
  const answer = String(info[qtype]);
  return { cidr, qtype, question: QUESTION_TEXT[qtype](cidr), answer };
}

function runAuto(numQuestions = 5) {
  for (let i = 1; i <= numQuestions; i++) {
    const p = generateProblem();
    console.log(`Q${i}. ${p.question}`);
    console.log(`    Answer: ${p.answer}`);
  }
}

function runInteractive(numQuestions = 10) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  let score = 0;
  let i = 0;

  function next() {
    if (i >= numQuestions) {
      console.log(`\nFinal score: ${score}/${numQuestions}`);
      rl.close();
      return;
    }
    i++;
    const p = generateProblem();
    rl.question(`\nQ${i}. ${p.question}\n> `, (answer) => {
      if (answer.trim() === p.answer) {
        console.log("Correct.");
        score++;
      } else {
        console.log(`Incorrect. Correct answer: ${p.answer}`);
      }
      next();
    });
  }
  next();
}

const args = process.argv.slice(2);
if (args[0] === "--auto") {
  runAuto(Number(args[1]) || 5);
} else {
  runInteractive(Number(args[0]) || 10);
}
